wicker/core/storage.py

```diff
@@ -44,27 +48,65 @@ def fetch_file(self, input_path: str, local_prefix: str, timeout_seconds: int) -
         """
         pass
 
+    @abstractmethod
+    def put_file(self, input_path: str, target_path: str) -> None:
+        """Put file on data storage in target location.
+
+        :param input_path: input path of file
+        :type input_path: str
+        :param target_path: target path to place file
+        :type target_path: str
+        """
+        pass
+
+    @abstractmethod
+    def put_object(self, object_bytes: bytes, target_path: str) -> None:
+        """Put object on data storage in target location.
+
+        :param object_bytes: Bytes of object to store.
+        :type object_bytes: bytes
+        :param target_path: Path to storage destination.
+        :type target_path: str
+        """
+        pass
+
+    def _safe_file_download(self, input_path: str, local_dst_path: str) -> str:
+        """Download file safely from data source with locking.
+
+        Locks the file path and downloads the file to the local system. Does not release the lock
+        until the file is downloaded fully.
+
+        Args:
+            input_path (str): path to file to download
+            local_dst_path (str): path to save file at on local system
+
+        Returns:
+            str: local path of the file when downloaded
+        """
+        os.makedirs(os.path.dirname(local_dst_path), exist_ok=True)
+        lock_path = local_dst_path + ".lock"
+        success_marker = local_dst_path + ".success"
+        while not os.path.isfile(success_marker):
+            with SimpleUnixFileLock(lock_path, timeout_seconds=self.read_timeout):
+                if not os.path.isfile(success_marker):
+                    filedir = os.path.split(local_dst_path)[0]
+                    os.makedirs(filedir, exist_ok=True)
+                    try:
+                        shutil.copyfile(input_path, local_dst_path)
+                        with open(success_marker, "w"):
+                            pass
+                    except Exception:
+                        logging.error(f"Failed to download/move object for file path: {input_path}")
+                        raise
+        return local_dst_path
+
```

Comment:
@aalavian there will be a little bit of code duplication here with the function `fetch_file` in the `S3DataStorage` class. I think it is ok as the actual calls inside the two functions differ a bit (one calls `shutil.copyfile` and the other downloads from S3) and it is somewhat awkward to resolve in a common function.

Another major consideration is that the `fetch_file` function in `S3DataStorage` is critical to Wicker performance in downloading from S3. If it's ok, I'd like to leave that function completely alone (rather than refactoring it to remove the code that is duplicated above) in order to eliminate any chance of accidentally damaging how the function operates.